### ISSUE TITLE
InMemoryResourceMatcher AND resource search parameters are evaluated as OR parameters

### ISSUE THREAD

**Issue by jarimayenburg** [hapifhir/hapi-fhir#2366]
**Describe the bug**
The `InMemoryResourceMatcher` contains a bug where AND search parameters are evaluated and matched as OR parameters.

This causes issues where for example `GET /fhir/DocumentReference?indexed=ge2018-01-01&indexed=le2019-01-01` would return DocumentReferences with an indexed data of 2020-01-01. This of course only happens when search parameter matching is done using the `InMemoryResourceMatcher` in the resource provider.

**To Reproduce**

For example:
1. Implement a DocumentReference resource provider which searches for DocumentReferences using the `InMemoryResourceMatcher`.  Support "indexed" as a search parameter.
2. Take for example the following DocumentReference as its only source resource to match incoming requests on:
```
{
  "resourceType": "DocumentReference",
  "status": "current",
  "type": {
    "coding": [
      {
        "system": "http://loinc.org/",
        "code": "68688-1",
        "display": "Neurological surgery Discharge summary"
      }
    ]
  },
  "indexed": "2020-01-01T00:00:00+01:00",
  "content": [
    {
      "attachment": {
        "contentType": "application/pdf",
        "language": "en",
        "url": "Binary/1234"
      }
    }
  ]
}
```
3. If you now perform the request `GET /fhir/DocumentReference?indexed=ge2018-01-01&indexed=le2019-01-01` it should return an empty Bundle, but it matches the above DocumentReference.

**Expected behavior**
It should not match the above DocumentReference

**Environment (please complete the following information):**
 - HAPI FHIR Version: Tested in 5.1, but should happen in 5.2 as well

**Additional context**
I've implemented a fix for this issue in #2365. I think it's good to implement some tests before we merge that one, so that we can be sure this issue actually exists and if it's indeed fixed by that PR.

**Comment by jarimayenburg** [comment]
I've written a test and included it in PR #2367. This test will fail until #2365 is merged. It's also included in the branch of #2365, where it succeeds.

**Comment by jamesagnew** [comment]
Fix has been merged
